BPF: Encrypted Overlay Support

ldelossa opened the pull request:
This pull request implements the data-path side of "Encrypted Overlay" support. 

"Encrypted Overlay" allows Cilium to IPSec encrypt overlay traffic and deliver the encrypted overlay traffic to a remote node via IPSec tunnel. 

A subsequent PR will follow which enables this in the IPSec control plane, creating the necessary policies and states to configure the tunnel source and destination addresses correctly. 

This feature is useful when an application's traffic destined for a remote host cannot be routed into the XFRM netfilter hooks for various reasons. 
In these cases we can first encapsulate the traffic in VXLAN, catch it at the egress interface, and recirculate the packet back into the system, at which point the traffic looks like unicast traffic that should be routed off host (and XFRM will be applied at this point). 

It's important to note that the following sysctl configs must be applied to any interfaces which tunnels traffic toward a remote node for this to work
```
- net.ipv4.conf.default.rp_filter = 0
- net.ipv4.conf.default.accept_local = 1
```

```release-note
bpf: introduce encrypted overlay datapath support 
```

ldelossa committed:
bpf: encrypt overlay traffic

This commit introduces the ability to encrypt overlay traffic before it
leaves the host.

The 'cil_to_netdev' function is updated to sniff into overlay packets
(only VXLAN supported for now) and determine if the ENCRYPTED_OVERLAY_ID
security identifier is present in the overlay's header.

If it is, a new function in encrypt.h will set the appropriate packet
mark on the skb and redirect the packet to the ingress of the interface
it was egressing on.

When the packet is seen on the ingress side of the device it will be
submitted to the XFRM hooks in the output routing path and the XFRM
subsystem will encrypt the packet.

Subsequent changes to the IPSec control plane to create the appropriate
states and policies to support this are required.

Signed-off-by: ldelossa <louis.delos@isovalent.com>

ldelossa committed:
bpf: add trace notification for overlay encryption

Add a trace notification when we are redirecting a packet back into the
stack for XFRM encryption.

Trace example:
-> stack flow 0xc218244b , identity unknown->unknown state encrypt-overlay ifindex 0 orig-ip 0.0.0.0: 172.18.0.3:58167 -> 172.18.0.2:8472 udp

Signed-off-by: ldelossa <louis.delos@isovalent.com>

ldelossa committed:
bpf,test: add tests for vxlan helper functions

Add unit tests for new vxlan helper functions in tunnel.h

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa commented on bpf/bpf_host.c:
```diff
@@ -1345,15 +1380,15 @@ int cil_to_netdev(struct __ctx_buff *ctx __maybe_unused)
 	}
 #endif
 
+	if (!validate_ethertype(ctx, &proto)) {
```
This would have occurred 3 separate times in different header guards, so I hoisted it up and made it evaluate unconditionally, I think the code is cleaner this way.

joestringer replied:
Minor nit here but there's a common style in C (particularly Linux C) where you try to keep all the return statements around the same place at the end of any long function, mainly to ensure that when the function returns a set of consistent things occur (for example, either send a drop notification or trace notification and then return). Then earlier in the function all exit paths just do a jump to 'out' or 'exit'.

I note that this function doesn't follow that style though, with inconsistent usage of either direct returns or jumping to the exit label, so no point litigating this.. Looking through I do notice however that some of the bare `return` statements are missing `send_trace_notify()` calls. This is the sort of inconsistency that the style is intended to avoid (or at least document the intent clearly with a comment, if there's good reason to skip generating events based on some specific condition).

ldelossa replied:
@joestringer this particular block of code has not changed from the previous commit, just moved. 

If we want to do refactoring of this block, I'd be happy too, but maybe we do this in a dedicated PR which addresses this concern directly?

ldelossa replied:
Going to resolve this as its purely cosmetic and this change is carried over from the previous code. Happy to address this in some other housekeeping commits outside of this PR tho.

joestringer replied:
Yep makes sense, agree it's already the case with this code and doesn't need to be implemented here in this PR.

brb reviewed (approved):
@cilium/wireguard changes LGTM.

pchaigno commented on bpf/bpf_host.c:
```diff
@@ -1292,26 +1292,61 @@ int cil_from_host(struct __ctx_buff *ctx)
+		return ret;
+
+	if (tunnel_vxlan_get_vni(data, data_end, ipv4) ==
+				 MARK_MAGIC_ENCRYPT) {
```
So the VNI will be completely replaced to only contain the magic bits and not the source security identity anymore? How will you retrieve the source secID?

ldelossa replied:
Solid point. 

The reason I opted to check the encrypt bits in isolation is because we are checking the mark in a new code path. I was not sure if say, this code path did not expect to ever see the encrypt bits set, and by **chance** a different mark comes in with the encrypt bits set (along with others) but without the intention for encryption (we use marks alot lol).

That being said, if we can work out that this 'over matching' of the mark cannot occur or simply won't, this can change to a '&' where we just mask-check the encryption bits.

pchaigno replied:
It's hackish but I think we can use the 8 cluster ID bits of the identity in the mark to set the encrypt bit. That way we retain the source security identity. The only case where we might actually need the 8 cluster ID bits for the identity is if Cilium Mesh is enabled (as discussed with Martynas on Slack). So we would need to make multisec + Cilium Mesh incompatible.

The alternative is to get the inner destination IP address and use it to lookup the SPI in the ipcache.

ldelossa replied:
Ugh, I would hate to cripple cluster mesh support for this right off the bat. 

But let's explore that path a bit.
When cluster mesh is disabled, do we still retain the 8 bits in the security ID as reserved? I see a bunch of code like this:
```
static __always_inline __u32
extract_cluster_id_from_identity(__u32 identity)
{
	return (__u32)(identity >> IDENTITY_LEN);
}
``` 
~~Making me feel like like the ratio of identity to cluser_id bits is variable via config? Not sure if this means that when cluster mesh is disabled, sec_id's become a full 32bits?~~

Later realized sec_id size is pinned to vxlan.vni size.

joestringer replied:
> The alternative is to get the inner destination IP address and use it to lookup the SPI in the ipcache.

^ This seems viable.

If we combine this with the following then I think we can mitigate some of the impacts:
1. Use a reserved identity (for example, `11`) and not `MARK_MAGIC_ENCRYPT` / `0x0E00`, which is a cluster-scoped identity, then we can avoid overloading that security identity value. We could guarantee that we never expect VNI `11` to be legitimately used as a security identity for policy enforcement purposes.
2. Ensure the datapath replaces this reserved identity with the real identity, and drop traffic that attempts to leave the node with this reserved identity (and potentially drop any traffic that comes into the node with that VNI).

ldelossa replied:
I want to confirm the approach I'm thinking here, since Paul's exact words:

> The alternative is to get the inner destination IP address and use it to lookup the SPI in the ipcache.

Are not the same as the solution we were discussing. 

What I'm thinking: 

When we get to `do_encrypt_overlay` function in this PR and, after we confirm that `vxlan.vni == 11`, we can replace the VNI. 

At this point, we will sniff into the vxlan packet to identify the **source** address of the encapsulated packet, perform a local endpoint map lookup for the source pod, and rewrite the VNI field with the discovered security ID, set the encryption **mark** and then redirect it. 

I'm not sure if you have yet another idea here Paul involving grabbing the SPI by sniffing at the destination of the encap'd packet?

pchaigno replied:
To get the SPI we need to lookup the inner destination IP, but we can probably get the SPI via the updated node ID map as mentioned in the other thread. To get the source security identity, you need to lookup the inner source IP. So Joe's solution makes sense to me here.

Note that the second approach I proposed, which avoids parsing VXLAN and doing an ipcache lookup, is incompatible with Cilium Mesh, not with clustermesh.

ldelossa replied:
> To get the SPI we need to lookup the inner destination IP, but we can probably get the SPI via the updated node ID map as mentioned in the other thread.

valid point here. 

Since we are now heading towards sniffing into the vxlan packet, we could just feed the dest to find the SPI as well... Which would mean less code changes and an alternative to placing it in NodeID map :thinking: 

> So Joe's solution makes sense to me here.

:+1:

> Note that the second approach I proposed, which avoids parsing VXLAN and doing an ipcache lookup, is incompatible with Cilium Mesh, not with clustermesh.

Ugh sorry, I confused my meshes :(

pchaigno replied:
> Since we are now heading towards sniffing into the vxlan packet, we could just feed the dest to find the SPI as well... Which would mean less code changes and an alternative to placing it in NodeID map 🤔

Yep. It's one additional LPM lookup vs. a hashmap lookup, but I'm not sure it makes a big difference considering all of this is encrypted anyway.

ldelossa replied:
Yes, I'm leaning toward doing this, then this PR is completely self contained and can find SPIs correctly. 

We can always keep the idea of moving SPI into NodeID map as a feature enhancement if this is still desired from an overall IPSec design.

ldelossa replied:
After some more thought Paul, we STILL may have situations where we are trying to get some traffic not related to a destination security ID over to the other node. 

Therefore, we still need to do SPI lookup via the InternalNodeIP or CiliumNodeIP. It probably makes more sense to just do this in the NodeID map after all...

ldelossa replied:
This should be fixed with latest push. Unresolve if you feel otherwise

pchaigno commented on bpf/lib/encrypt.h:
```diff
@@ -132,6 +133,69 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+	if (!info)
+		return DROP_INVALID;
+
+	encrypt_key = get_min_encrypt_key(info->key);
```
Won't `info->key` always be zero given you did the ipcache lookup with the outer IP address, so a node IP address?

ldelossa replied:
Ah, you're right, hmm.

So we talked about how we can add SPI into the NodeID map, which allows us to retrieve an SPI for a node's CiliumInternalIP and NodeInternalIP. 

That's not implemented yet, but when it is, we could do a query of this before `	encrypt_key = get_min_encrypt_key(info->key);` to get the SPI and then feed this into the function instead of info->key

pchaigno replied:
Ah, yes. Let's table this until then.

pchaigno commented on bpf/lib/encrypt.h:
```diff
@@ -132,6 +133,69 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+	 * daddr of the packet to this function
+	 */
+	ret = set_ipsec_encrypt(ctx, encrypt_key, ip4->daddr,
+				info->sec_identity, false);
```
Won't `info->sec_security` be the security identity of the source node instead of the source pod?

ldelossa replied:
Because we are sending 'encrypted vxlan' I was thinking, the 'sender' actually becomes the host. This is because the host is wrapping the vxlan into ESP and sending it over. Maybe this is not the right perspective?

pchaigno replied:
This is the identity that is going to be used for policy enforcement on the other side, at pod ingress. If users receive multicast packets from a pod frontendA won't they want to allow that pod rather than its host (and all multicast traffic sent from pods on that host)?

ldelossa replied:
This information is still available after ESP and VXLan is decap'd, do we optimize and not do a ipcache lookup on the source, on the remote node, before policy ?

If not, then this can change to the source ID. TBH, multicast hasn't had any policy applied to it yet, so I'm most likely overlooking this.

ldelossa replied:
Okay after a bit of discussion out of band, doing an IPCache lookup at the remote because the VNI is not valid is a complete no-go. So we can ignore my comment above.

ldelossa replied:
This should be fixed with latest push. Unresolve if you feel otherwise

pchaigno commented on bpf/lib/trace.h:
```diff
@@ -54,6 +54,7 @@ enum trace_reason {
 	TRACE_REASON_UNKNOWN,
 	TRACE_REASON_SRV6_ENCAP,
 	TRACE_REASON_SRV6_DECAP,
+	TRACE_REASON_ENCRYPT_OVERLAY,
```
It seems we've started to misuse this enum because we misnamed it. It shows as `state encrypt-overlay` in the traces :confused: We should probably discuss what we want to do here. Maybe just fix display in `cilium monitor` and Hubble?

ldelossa replied:
Hmm, what was the original intent for 'state' ? 
This is a pretty handy spot to just explain 'why' this tracepoint was hit.

kaworu replied:
> Maybe just fix display in cilium monitor and Hubble?

Yes that make sense to me. I think the idea of `trace_reason` being correlated with state doesn't hold anymore since non-CT trace reasons were introduced.

For Hubble there is nothing to update since we don't expose (yet) the trace reason directly (Hubble only infer some other fields from it).

For `cilium monitor` [`TraceNotify.DumpInfo`](https://github.com/cilium/cilium/blob/4082bc38407a3fd2b59f2bd1e8e0e1ed06be8340/pkg/monitor/datapath_trace.go#L274) should be updated and maybe [`TraceNotifyVerbose`](https://github.com/cilium/cilium/blob/4082bc38407a3fd2b59f2bd1e8e0e1ed06be8340/pkg/monitor/datapath_trace.go#L346-L362) depending on how willing we are to break `cilium monitor --json` compatibility.

ldelossa replied:
Breaking the json compat is a really good question. I think we need to bring this up in community chat/meeting.

learnitall replied:
Have we had a chance to bring this up yet?

ldelossa replied:
@learnitall nope.

joestringer replied:
Why do we need a tracepoint for this at all?

Originally we had effectively "packet received" and "packet transmitted" trace events, which generally help the user to understand where packets are flowing. We also then used this to generate things like Hubble Service Map to demonstrate the way packets are flowing between Endpoints.

If we wanted to debug something, for instance because the packets are mishandled, then typically we would enable debug and then have various dedicated debug messages that can go into much more detail.

We're hitting issues in the project at this point where the datapath generates too many events because we aren't very careful about when we emit events. I can see the value in being able to tell whether the packet was encrypted or not, but why is this particular feature so different from the various other features? Why does it require another dedicated message per packet handled by this specific code? Ultimately Cilium is delivering or not delivering the packets, then maybe there's some extra properties (why, how). In an ideal world I might suggest that we just collect a set of bits like "tunneled", "encrypted", etc. and then merge those into the final "packet transmitted" event. Though I understand that feeding this information to the right place in the datapath can be a bit tricky with the current structure.

Maybe to reframe the question: Could we get away with a debug message here instead of trace?

joestringer replied:
Revisiting after seeing the individual commit with the context from the git commit message:

```
    bpf: add trace notification for overlay encryption

    Add a trace notification when we are redirecting a packet back into the
    stack for XFRM encryption.

    Trace example:
    -> stack flow 0xc218244b , identity unknown->unknown state encrypt-overlay ifindex 0 orig-ip 0.0.0.0: 172.18.0.3:58167 -> 172.18.0.2:8472 udp
```

Seeing `-> stack` for a message like this actually makes some sense to me: Basically the packet is being redirected back `to stack`, so it's easier to follow the flow of the packets. The downside is that this means we could end up sending at least two (maybe three?) `TRACE_TO_*` messages per packet, which seems a bit extraneous and may have an adverse impact on CPU usage with Hubble. If I follow correctly though this will only happen if the user lowers the monitor aggregation level below medium. Is that correct?

Looking at `state encrypt-overlay`, I agree it seems a little bit of an odd UX representation. I think the only reasons we emitted trace notifications in a much earlier version was in order to understand the connection tracking state, hence the "reason" field in the datapath only contained CT-based reasons and the CLI presented these as "state". Not sure what Hubble does here. But I agree we can solve this on the CLI UX.

ldelossa replied:
> If I follow correctly though this will only happen if the user lowers the monitor aggregation level below medium. Is that correct?

Yes that is the case.

ldelossa replied:
Judging from the comments here, and that Alex has already opened an issue for handling the state a bit better, I'm going to resolve this. If there' still an issue here feel free to unresolve and ping me.

kaworu replied:
> Not sure what Hubble does here. But I agree we can solve this on the CLI UX.

Hubble doesn't expose the trace reason yet (see https://github.com/cilium/cilium/pull/31226).

> Alex has already opened an issue for handling the state a bit better

If you're referring to https://github.com/cilium/cilium/pull/31226, the PR doesn't change the `cilium monitor` output, so as far as I understand "Maybe just fix display in cilium monitor" is still an open question (see [my previous comment](https://github.com/cilium/cilium/pull/31073#discussion_r1509109408)).

pchaigno reviewed (commented):
Neat pull request! I really like the code comments :slightly_smiling_face: 

Couple comments below. IIUC, I think there's an issue with the SPI and security identity.

kaworu reviewed (changes requested):
Thanks for the PR @ldelossa, on the Hubble side:
1. Please add `TraceReasonEncryptOverlay` introduced by this PR in [`TraceReasonIsEncap`](https://github.com/cilium/cilium/blob/4082bc38407a3fd2b59f2bd1e8e0e1ed06be8340/pkg/monitor/datapath_trace.go#L163-L170). That way the corresponding Hubble flow won't have [the `IsReply` field](https://github.com/cilium/cilium/blob/4082bc38407a3fd2b59f2bd1e8e0e1ed06be8340/api/v1/flow/flow.proto#L84-L87) set. The test case should be similar to [SRV6 encap](https://github.com/cilium/cilium/blob/4082bc38407a3fd2b59f2bd1e8e0e1ed06be8340/pkg/hubble/parser/threefour/parser_test.go#L794-L803), except that the `Source` should be `hostEP` instead (since this trace is emitted from `bpf_host.c`).
2. I'm not sure whether [`decodeTrafficDirection`](https://github.com/cilium/cilium/blob/4082bc38407a3fd2b59f2bd1e8e0e1ed06be8340/pkg/hubble/parser/threefour/parser.go#L459) will correctly infer EGRESS for the corresponding flow. Can you run `hubble observe -o json` and see what the `TraceReasonCtReply` flow looks like? If shown as INGRESS, please add a comment in `decodeTrafficDirection` stating that it is currently broken for this particular trace reason :pray:

joestringer commented on bpf/bpf_host.c:
```diff
@@ -1391,6 +1426,24 @@ int cil_to_netdev(struct __ctx_buff *ctx __maybe_unused)
+		/* we are redirecting back into the stack, so TRACE_TO_STACK
+		 * for tracepoint
+		 */
+		send_trace_notify(ctx, TRACE_TO_STACK, 0, 0, 0,
+				  0, TRACE_REASON_ENCRYPT_OVERLAY, 0);
```
Two points here:
- I don't know if you've considered the meaning of all these `0`s, but I suspect that we're avoiding providing debugging information about the flow here, which will make this harder to debug if a user tries to figure out what's going on based on this trace message. Ideally very few of these `0`s should be `0`s, we can provide the real information about the packet at this point in time.
- I think we need to consider how many of these messages we want to generate and how this integrates with `monitor-aggregation`. If this is part of encrypt hotpath then it could be emitted for every packet, forcing Cilium to spend a bunch of CPU processing these events.

ldelossa replied:
This maybe a bit of a unique case, since when we get here we have vxlan encap'd traffic already. 

Given our last convo, we are going to have to pull the source information out to rewrite the VNI, so we can fill in the src ID, however, I'm not sure if filling in the dest_id is worth the effort, as we'd need to sniff into the vxlan traffic for the dest and do an additional map lookup. 

re: `monitor-aggregation` its a good question, this tracepoint is right before we redirect the packet back into the interface and route it toward XFRM so it can do the encryption. What is our stance for traces like this? Aren't all our traces susceptible to hot paths if the application starts bursting data?

joestringer replied:
> Aren't all our traces susceptible to hot paths if the application starts bursting data?

Yep, and it's already a problem for CPU usage at high packet rates; I wouldn't want to extend this further with additional trace events that have no way to opt-out. There's been some PRs recently to try to tackle some of this, so maybe it's worth poking around in the recent hubble/monitor PRs and raise some discussion about the right way to introduce this visibility while accounting for the cost (or allowing it to be disabled).

pchaigno replied:
What Louis did here is hide those traces whenever monitor aggregation based on CT is enabled (cf. [`lib/trace.h`](https://github.com/cilium/cilium/blob/6b98a0b210ea2eae59e2eeb399aad3ce121f3caf/bpf/lib/trace.h#L174-L181)). So by default those traces won't be emitted. It matches Ryan's recent changes in https://github.com/cilium/cilium/pull/31015/commits/5fc9dd75d42fcf9214ebb8bd408ee91063f3c865 and https://github.com/cilium/cilium/pull/29616/commits/897b8edd9c495b29cc3344fd7fa300adda4b1b5b.

Not to say we shouldn't fill the contextual information (ex. security identity) if we have it.

joestringer replied:
Something else I didn't notice the first time... is that every other `CTX_ACT_REDIRECT` path in this function *avoids* calling `send_trace_notify()`. I don't know if there's a specific reason for this or if it's an oversight, because the code structure doesn't make this clear. Maybe that's just a general datapath cleanup thing to follow up on.

ldelossa replied:
Yup make sense. 

I personally feel like a trace notification is really valuable here. The packet-path for encrypted overlay is kinda complex IMO, so knowing that traces which show the major leg of the flow is important IMO.

ldelossa replied:
I think we landed on keeping the trace as it makes sense since we are handing the packet off to the stack. So going to resolve. If this is not the case please unresolve and ping me.

ldelossa committed:
datapath: add EncryptedOverlayID reserved ID 11

This commit adds a new reserved security identity for signaling overlay
traffic which must be IPSec encrypted.

When the eBPF datapath encounters an egress packet with this security
identity in an overlay header (currently only VXLan supported) it will
subject the packet to IPSec encryption and rewrite the overlay header
with the correct security identity before the packet leaves the host.

Therefore, this identity should NEVER be seen on traffic ingress or
egress the node from the network.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

kaworu commented on bpf/lib/encrypt.h:
```diff
@@ -132,6 +133,82 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+	/*
+	 * this is a vxlan packet so ip4->daddr is the tunnel endpoint
+	 */
+	ret = set_ipsec_encrypt(ctx, encrypt_key, ip4->daddr, ep_info->sec_id,
```
`encrypt_key` is always zero here right?

ldelossa replied:
Yes. Until we implement getting SPI from NodeID map.

kaworu commented on pkg/hubble/parser/threefour/parser_test.go:
```diff
@@ -802,6 +802,17 @@ func TestDecodeIsReply(t *testing.T) {
+	// TRACE_TO_STACK Encrypted Overlay
+	tn = monitor.TraceNotifyV0{
+		Type:     byte(monitorAPI.MessageTypeTrace),
+		Source:   localEP,
```
I think this needs to be
```suggestion
		Source:   hostEP,
```
Since the trace is emitted by `bpf_host.c`

ldelossa replied:
@kaworu sounds good to me, but should the test have failed? It passes like this.

kaworu replied:
No I think it's expect too pass either way, it's was more like a cleanup request :pray:

kaworu commented on pkg/monitor/datapath_trace.go:
```diff
@@ -163,6 +165,8 @@ func TraceReasonIsKnown(reason uint8) bool {
 	switch reason {
 	case TraceReasonSRv6Encap:
+		fallthrough
+	case TraceReasonEncryptOverlay:
```
nit:
```suggestion
	case TraceReasonSRv6Encap, TraceReasonEncryptOverlay:
```

joestringer commented on bpf/lib/mcast.h:
```diff
@@ -341,6 +342,7 @@ static long __mcast_ep_delivery(__maybe_unused void *sub_map,
 				struct _mcast_ep_delivery_ctx *cb_ctx)
 {
 	int ret = 0;
+	__u32 tunnel_id = 2;
```
nit: If this is `WORLD_ID`, it should be `WORLD_ID`, not `2` :sweat_smile: 
```suggestion
	__u32 tunnel_id = WORLD_ID;
```

ldelossa replied:
Ack, this was there in the original PR too, but I guess it does not hurt to update in this diff.

joestringer commented on bpf/lib/wireguard.h:
```diff
@@ -15,19 +15,15 @@
-	if (!validate_ethertype(ctx, &proto))
-		return DROP_UNSUPPORTED_L2;
-
 	switch (proto) {
```
I'm a little surprised to see what was previously considered "unsupported L2" to now _not_ be dropped and actually pass somewhere in this patch. Have the implications of this been explored? Should this fragment be part of a separate patch with more details in a commit message about the motivation for this change and why we think it's safe / won't leak L2 packets?

Is this trying to facilitate ARP or something on this interface? I thought ESP typically occurs after IP headers, so we shouldn't need to allow arbitrary L2 at this point in order to allow IPsec packets to pass here.

joestringer replied:
Oh, I didn't realize that this was part of the refactor to check the ethertype much earlier. I'd suggest pulling these changes out into their own patch if possible, keep them separate from this patch. Makes it easier to confirm that this is a no-op change.

ldelossa replied:
This still occurs, but at the caller, this particular function is only called in one place and was duplicating 'prep' logic which was occurring in the caller. This was checked out by @brb and approved. 

I did confirm that there is no other possible path to this function, and for the only one there is, this l2 check occurs unconditionally.

ldelossa replied:
This is resolved by pulling out the refactor bits into their own commit. 

Unresolve and ping me if this is not a satisfactory resolution.

joestringer commented on pkg/hubble/parser/threefour/parser.go:
```diff
@@ -481,6 +481,7 @@ func decodeTrafficDirection(srcEP uint32, dn *monitor.DropNotify, tn *monitor.Tr
 
 			// isSourceEP != isReply ==
 			//  (isSourceEP && !isReply) || (!isSourceEP && isReply)
+			// FIXME: currently broken for monitor.TraceReasonEncryptOverlay showing INGRESS
```
Would you mind filing an issue for this? Elsewhere in Cilium we often use the format like:

```suggestion
			// GH-XXXX: currently broken for monitor.TraceReasonEncryptOverlay showing INGRESS
```

This just makes it easier to (a) track the fix, (b) attract someone to fix it, (c) conduct any discussion about solutions.

joestringer replied:
Actually the discussion point I raised immediately below may render this point unnecessary.

Something else I note though is that `TraceReasonEncryptOverlay` is _probably_ showing up as INGRESS because the packet is INGRESSing the stack right? I don't know if we've 100% locked down the meaning of INGRESS/EGRESS but if this is effectively the Linux stack perspective, ie TC ingress/egress then it may even be accurate. Confusing for users, but somewhat consistent with existing flows.

ldelossa replied:
@kaworu I think we are already tracking an issue for this, no?  

I made the same comment to Alex. Ingress is actually correct here, but it may not be correct or appropriate when the flow is user facing in hubble. 

I think this should become more of a conversation here: https://github.com/cilium/cilium/pull/31226 or in a follow up issue as I don't believe these semantics are really a focus of this PR.

ldelossa replied:
Leaving this open until we determine the right gh-ticket to add to the comment above.

kaworu replied:
> @kaworu I think we are already tracking an issue for this, no?

No issue yet, I'm intending to address it as part of https://github.com/cilium/cilium/pull/31226.

> [...] but somewhat consistent with existing flows.

@joestringer As you mentioned in another thread on this PR, the packet will be processed later and generate another Hubble flow that will be shown as EGRESS (which at this point will make complete sense). For this trace notification, while INGRESS can be seen technically correct, I don't think having the same packet generating an INGRESS _and_ an EGRESS flow make sense overall at a high level (e.g. metrics accounting). Thus my plan is basically to consistently set `traffic_direction=EGRESS` on Hubble flows generated from a `TraceReasonEncryptOverlay` notification.

ldelossa replied:
Cool, at the minimum I will update the comment in the code to point to #31226 :)

ldelossa replied:
Comment updated, resolving.

joestringer commented on bpf/bpf_host.c:
```diff
@@ -1298,12 +1298,46 @@ int cil_from_host(struct __ctx_buff *ctx)
+ *
+ * IS_ERR can be used to determine if this function ran into an error.
+ */
+static __always_inline int do_encrypt_overlay(struct __ctx_buff *ctx,
```
nit:
```suggestion
static __always_inline int maybe_encrypt_overlay(struct __ctx_buff *ctx,
```

joestringer replied:
^ I would kinda expect the `do_foo()` to actually do the foo. If there's a layer which does input validation then another one than actually does the thing, then the outer layer should be either `foo()` or `maybe_foo()` and then the inner one should be the `do_foo()`

ldelossa replied:
Ah, I never liked the 'maybe' wordage since the return value should really tell you 'what' happened. But if this is preferred or a common semantic I can change it.

joestringer replied:
:+1: I guess this is maybe more preference than anything then.

When I compared with `do_decrypt()` which steals the packet away, I found it slightly surprising that actually only some specific vxlan traffic with `ENCRYPTED_OVERLAY_ID` will be handed by this function; `do_encrypt_overlay()` seems pretty decisive/final. In the back of my mind I can also envision the next developer to come along and inspect this code to think that this is the primary ipsec encryption path.. but that said, I don't think a function rename will help with that.

I'll resolve this, doesn't seem particularly notable to resolve one way or another.

joestringer commented on bpf/lib/encrypt.h:
```diff
@@ -132,6 +133,82 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+ * Be aware that the redirected-to interface needs to have the following
+ * sysctl enabled for this to work correctly (per-device is fine)
+ *   - net.ipv4.conf.default.rp_filter = 0
+ *   - net.ipv4.conf.default.accept_local = 1
```
May be worth reviewing exactly what sort of traffic this may allow _other_ than the traffic intended in this PR. I'm not sure I fully understand the consequences of doing this. I agree that in order to implement this feature the way it's implemented, this is necessary... but it may have other impacts as well.

From kernel docs:

> accept_local - BOOLEAN
>	Accept packets with local source addresses. In combination with
>	suitable routing, this can be used to direct packets between two
>	local interfaces over the wire and have them accepted properly.
>	default FALSE

So then traffic from outside the node (for instance from MitM attacker) could be generated with locally assigned addresses, and the node will accept & process those packets rather than immediately dropping them.

If we're not comfortable with this, we could consider some rewrite of the local address before forwarding so that it bypasses this check, then restore it on the final transmit. But I'm probably jumping the gun by suggesting such changes without fully understanding the implications of what the change really means.

joestringer replied:
There may be another mitigation for this by ensuring the device for redirection here is distinct from the native routing device (for example by reusing `cilium_host` or `cilium_net`). That way we could have higher confidence that the only actor that can place packets there with local addresses is this specific path / BPF prog.

ldelossa replied:
> If we're not comfortable with this, we could consider some rewrite of the local address before forwarding so that it bypasses this check, then restore it on the final transmit. But I'm probably jumping the gun by suggesting such changes without fully understanding the implications of what the change really means.

I think adding any more 'translation' to this feature is a no-go. 

> There may be another mitigation for this by ensuring the device for redirection here is distinct from the native routing device (for example by reusing cilium_host or cilium_net).

I considered this, but this adds even **more** redirection to this feature which I wasn't really comfortable with. We'd be routing across quite a few interfaces and doing double redirection in these cases. I'm not sure if its worth it. 

I think we need to communicate that these sysctls are required and the user can make a determination of whether this is OK or not.

joestringer replied:
As long as that's not a deal-breaker for users then that's fine. I'm not sure I fully understand the implications yet, so I can imagine users having some trouble understanding too.

ldelossa replied:
I do see your point. 

I'm going to add a follow-up point in our notes that we should return our attention to this. And see if we want the control plane handle this administration of the sysctl. 

I'd be a bit hesitant to transit through cilium_host, while it looks attractive at first, the flow becomes even more windy :\

ldelossa replied:
Resolving for now, I have made note about this topic and we can discuss further.

joestringer commented on bpf/lib/encrypt.h:
```diff
@@ -132,6 +133,82 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+ *
+ * Be aware that the redirected-to interface needs to have the following
+ * sysctl enabled for this to work correctly (per-device is fine)
+ *   - net.ipv4.conf.default.rp_filter = 0
```
Might be worth checking this against the `rp_filter` matches in the tree to see if this is already the case. Likewise rather than forcing users to set the setting (goes for `accept_local` too), we may want to configure these for them (assuming we have a good understanding of the implications and are happy to set these settings).

ldelossa replied:
These comments are just advising that these values must be on. Whether the control plane is doing this for you or not, is not really carried into this. This is just a place to say "hey if you're not getting here, check this out". If  we want the control plane to set the values, I think I can approach this in a follow up PR.

joestringer commented on bpf/lib/encrypt.h:
```diff
@@ -132,6 +133,82 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+	 * so redirecting to ingress side of the same interface doesn't fail.
+	 */
+	eth_load_saddr(ctx, &dst_mac, 0);
+	eth_store_daddr(ctx, &dst_mac, 0);
```
Idly wondering whether packets with identical src+dst mac might interfere with anything. Whether it'd be better to swap or come up with a new magic MAC for the new "source" since it's actually sourced locally.

ldelossa replied:
Was going to do a full swap, but then looking at the kernel  code for ingress, src mac didn't seem to really matter much at all on the ingress path, so I just did less work.

joestringer commented on bpf/lib/tunnel.h:
```diff
@@ -83,4 +88,135 @@ tunnel_vni_to_sec_identity(__be32 vni)
+	      + sizeof(struct udphdr));
+
+	old_vni = vx->vx_vni;
+	vx->vx_vni = bpf_htonl(vni << 8);
```
Why is the bitshift necessary?

I can understand clearing the upper 8 bits, but this is just straight up shifting it and then updating the header. :thinking:

joestringer replied:
Hmm, looking at `GetBPFValue()` in `pkg/maps/lxcmap/lxcmap.go` there's an inconsistency here. The userspace code comments say "store the sec_id in network byte order", but here the code is doing "host to network long" with the value from the map. Maybe part of the reason that the bitshift _seems_ necessary is because actually the VNI is already in NBO and no manipulation is actually necessary. This could potentially cause a platform bug on arm64.

ldelossa replied:
The integer value in `vni` must become a 24bit integer in big endian. 

We must left shift the integer representation such that integer is maintained but the bottom 8 bits are cleared (they are reserved in the vxlan header format). Then we flip he bits for endianness.

ldelossa replied:
> The userspace code comments say "store the sec_id in network byte order", but here the code is doing "host to network long" with the value from the map.

Huh, I never see a comparison to integers using ntoh like here:

```
static __always_inline bool identity_is_world_ipv6(__u32 identity)
{
#if defined ENABLE_IPV4 && defined ENABLE_IPV6
		return identity == WORLD_ID || identity == WORLD_IPV6_ID;
#else
		return identity == WORLD_ID;
#endif
}
``` 

But maybe this is because we know we are always a single byte value here?

ldelossa replied:
The usage of the function is here: 

```
	ep_info = __lookup_ip4_endpoint(inner_ipv4->saddr);
	if (!ep_info)
		return DROP_INVALID;

	/*
	 * this is a vxlan packet so ip4->daddr is the tunnel endpoint
	 */
	ret = set_ipsec_encrypt(ctx, encrypt_key, ip4->daddr, ep_info->sec_id,
				false);
	if (ret != CTX_ACT_OK)
		return ret;

	/*
	 * source mac is our current egress interface, lets copy it to dmac
	 * so redirecting to ingress side of the same interface doesn't fail.
	 */
	eth_load_saddr(ctx, &dst_mac, 0);
	eth_store_daddr(ctx, &dst_mac, 0);

	/* need to revalidate data since we just re-wrote mac addresses */
	if (!revalidate_data(ctx, &data, &data_end, &ip4))
		return DROP_INVALID;

	/* right now, the VNI of this packet is ENCRYPTED_OVERLAY_ID, we need
	 * to rewrite this VNI to the source's sec id before we transmit it
	 */
	if (!tunnel_vxlan_rewrite_vni(ctx, data, data_end, ip4,
				      ep_info->sec_id))
		return DROP_INVALID;
```

So ep_info->sec_id will be already big endian when I grab it out of the map?

joestringer replied:
After some out-of-band discussion, Louis and I concluded that the comment in the userspace lxcmap code is incorrect, lxcmap stores the Security Identity as a host byte-order u32. We can send a follow-up patch to drop that incorrect comment.

We also reasoned through how the bitshift accounts for the "reserved" bits on both little- and big-endian architectures, and particularly given the way that the header is declared in the tree with a __be32 and copied across directly in that form (retaining the lowest byte as 0 due to the reserved field). It's correct.

This logic also matches the equivalent/opposite function in `tunnel_vni_to_sec_identity()`.

ldelossa replied:
> This logic also matches the equivalent/opposite function in tunnel_vni_to_sec_identity(). 

Ah yes, very good point.

julianwiedmann commented on bpf/lib/tunnel.h:
```diff
@@ -83,4 +88,135 @@ tunnel_vni_to_sec_identity(__be32 vni)
+	hdr = (struct vxlanhdr *)(data + sizeof(struct ethhdr) + l3_size +
+	       sizeof(struct udphdr));
+
+	return bpf_ntohl(hdr->vx_vni) >> 8;
```
```suggestion
	return tunnel_vni_to_sec_identity(hdr->vx_vni);
```

ldelossa replied:
Resolved in latest push

julianwiedmann commented on bpf/lib/tunnel.h:
```diff
@@ -83,4 +88,135 @@ tunnel_vni_to_sec_identity(__be32 vni)
+				    BPF_F_PSEUDO_HDR | sizeof(__u16)) < 0)
+			return false;
+	}
+
```
Isn't this why `csum_l4_offset_and_flags()` sets `BPF_F_MARK_MANGLED_0` ?

ldelossa replied:
Yes, but if we know ahead of time, why even do the whole setup just to no-op?

julianwiedmann replied:
That *feels* like a pre-mature optimization, at the cost of being different from everyone else ;). But fair enough.

julianwiedmann commented on bpf/tests/tunnel_vxlan_helpers_tests.c:
```diff
@@ -0,0 +1,188 @@
+
+	vx->vx_vni = bpf_htonl(VXLAN_VNI << 8);
+
+	mk_data(encap_data);
```
How does this differ from `pktgen__push_ipv4_packet()` ?

ldelossa replied:
I tried this at first, but due to packet layer accounting, it did not work.

julianwiedmann replied:
Could you add a comment to that effect, so we know to clean it up when the layer-count gets bumped?

ldelossa replied:
Fixed in latest push

julianwiedmann commented on bpf/lib/tunnel.h:
```diff
@@ -83,4 +88,135 @@ tunnel_vni_to_sec_identity(__be32 vni)
+	if (ipv4->protocol != IPPROTO_UDP)
+		return false;
+
+	l3_size = ipv4->ihl * 4;
```
```suggestion
	l3_size = ipv4_hdrlen(ipv4);
```

(please check for other occurrences)

ldelossa replied:
Ugh, this issue I face here is including `#include "ipv4.h"` 
Then includes `dbg.h` and `map.h` and others and without following more include chains the compilation fails:

```
In file included from /home/louis/git/gopath/src/github.com/cilium/cilium-enterprise/bpf/lib/ipv4.h:10:
/home/louis/git/gopath/src/github.com/cilium/cilium-enterprise/bpf/lib/l4.h:60:10: error: use of undeclared identifier 'DROP_CSUM_L4'
   60 |                 return DROP_CSUM_L4;
      |                        ^
/home/louis/git/gopath/src/github.com/cilium/cilium-enterprise/bpf/lib/l4.h:63:10: error: use of undeclared identifier 'DROP_WRITE_ERROR'
   63 |                 return DROP_WRITE_ERROR;
      |                        ^
In file included from bpf_lxc.c:20:
In file included from ./lib/auth.h:7:
In file included from ./lib/common.h:21:
In file included from ./lib/tunnel.h:13:
In file included from /home/louis/git/gopath/src/github.com/cilium/cilium-enterprise/bpf/lib/ipv4.h:11:
In file included from /home/louis/git/gopath/src/github.com/cilium/cilium-enterprise/bpf/lib/metrics.h:13:
In file included from /home/louis/git/gopath/src/github.com/cilium/cilium-enterprise/bpf/lib/maps.h:8:
/home/louis/git/gopath/src/github.com/cilium/cilium-enterprise/bpf/lib/ipv6.h:54:11: error: use of undeclared identifier 'DROP_INVALID_EXTHDR'
   54 |                         return DROP_INVALID_EXTHDR;
      |                                ^
/home/louis/git/gopath/src/github.com/cilium/cilium-enterprise/bpf/lib/ipv6.h:57:11: error: use of undeclared identifier 'DROP_FRAG_NOSUPPORT'
   57 |                         return DROP_FRAG_NOSUPPORT;
      |                                ^
/home/louis/git/gopath/src/github.com/cilium/cilium-enterprise/bpf/lib/ipv6.h:64:12: error: use of undeclared identifier 'DROP_INVALID'
   64 |                                 return DROP_INVALID;
      |                                        ^
/home/louis/git/gopath/src/github.com/cilium/cilium-enterprise/bpf/lib/ipv6.h:81:9: error: use of undeclared identifier 'DROP_INVALID_EXTHDR'
   81 |         return DROP_INVALID_EXTHDR;
      |                ^
  ```
  
  If we want to have helper functions like this, they should really be within 'export only' headers (or at least headers which include other minimal headers), we have created a big nest of include declarations which I don't want to fix in this PR.

ldelossa replied:
Going to resolve this for now. I have a personal endeavor to make this situation better, once I unravel this stuff in a separate PR I will replace suggested line above (the suggestion is good).

julianwiedmann commented on bpf/lib/encrypt.h:
```diff
@@ -132,6 +133,82 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+	 * so redirecting to ingress side of the same interface doesn't fail.
+	 */
+	eth_load_saddr(ctx, &dst_mac, 0);
+	eth_store_daddr(ctx, &dst_mac, 0);
```
These can fail, no?

ldelossa replied:
Good catch, think I missed error handling here in haste. Will add this now.

ldelossa replied:
resolved in latest push

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -1308,12 +1308,46 @@ int cil_from_host(struct __ctx_buff *ctx)
+
+	/* only support vxlan for now */
+	if (TUNNEL_PROTOCOL != TUNNEL_PROTOCOL_VXLAN)
+		return ret;
```
`build_bug_on()` ?

ldelossa replied:
Could this get annoying when TUNNEL_MODE is on but TUNNEL_PROTOCOL is a protocol other then VXLAN? I'm worried that building tests or building in order to check for compilation issues may break.

pchaigno replied:
Not if all this code is behind a flag (e.g., `ENABLE_ENCRYPTED_OVERLAY`).

ldelossa replied:
I didn't add yet another 'ENABLE_ENCRYPTED_OVERLAY` and CLI flag because 
1. We have so many flags
2. The datapath needs to be hard-coded to opt-in to performing (I.E. by writing the ENCRYPTED_OVERLAY_ID identity into the tunnel_id). Therefore, I was thinking this feature just becomes part of IPSec and if certain parts of the datapath need its functionality, it will opt-in to it.

pchaigno replied:
I'm not following. How will we control if the overlay is encrypted or not? We don't want to enable this for existing users.

ldelossa replied:
Encrypted Overlay is 'opt-in' by the datapath right? Specific code needs to place the 'ENCRYPTED_OVERLAY_ID' in the tunnel_key.id to inform the DP that the traffic should be "encrypted overlay" traffic. Without this, nothing changes for any traffic in the cluster.

pchaigno replied:
OK, I think I understand now. You're saying that a consumer of encrypted overlay like multicast will know to set the new security ID. That's good, but wasn't the goal here to also enable the usual IPsec from using encrypted overlay? If we want to do that, I don't really see a way besides having a new flags (in agent and datapath).

julianwiedmann replied:
I think Paul is spot-on. There's no good reason why this code-path should be enabled by default - it adds BPF complexity, and exposes all existing users to potential bugs (that would otherwise only hit the multicast use-case).

ldelossa replied:
> That's good, but wasn't the goal here to also enable the usual IPsec from using encrypted overlay?

I'm not sure that's the case. I think the use-case here is to allow traffic which cannot enter XFRM hooks normally to have an alternative method to do so. I intended for this feature to run along-side normal IPSec encryption flow.

ldelossa replied:
Discussion with @pchaigno @julianwiedmann made it clear why a flag would be appropriate for this feature. 
I'll add this in with the next commit.

ldelossa replied:
Added a new 'encryption.ipsec.encryptedOverlay` (agent) and ENABLE_ENCRYPTED_OVERLAY (datapath) with latest push. Resolving.

joestringer reviewed (approved):
LGTM

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -1357,15 +1389,15 @@ int cil_to_netdev(struct __ctx_buff *ctx __maybe_unused)
+	}
+
 #ifdef ENABLE_HOST_FIREWALL
 	if (ctx_snat_done(ctx))
```
Sorry, I don't think is a change we can do. We would start dropping all sorts of traffic **by default**, for no good reason. Yes, it would most likely be *exotic* traffic - but it's still not the kind of behavioural change that a "encrypted overlay" PR should introduce.

ldelossa replied:
blahhh, you are right. I thought all code paths lead to this, but I guess I didn't account for traffic which is just 'passing through' to the host stack. I'll back this out.

ldelossa replied:
Okay this is backed out now and the change is pushed. Resolving.

julianwiedmann reviewed (changes requested):
Getting one blocker in before the ship takes off, sorry.

(still need to have an in-depth look at the other changes)

pchaigno commented on pkg/labels/labels.go:
```diff
@@ -51,6 +51,16 @@ const (
+	//
+	// This identity should never be seen on the wire and is used only on the
+	// local host.
+	IDNameEncryptedOverlay = "encrypted-overlay"
```
Is that the name that will show up in `cilium monitor`? Should it be `overlay-to-encrypt`?

ldelossa replied:
So far all the naming has been "encrypted overlay". Why the change now to 'overlay-to-encrypt' ?

pchaigno replied:
I'm just a bit worried it will look like the traffic is encrypted when looked at with `cilium monitor`, even though it isn't yet. I wasn't suggesting to change the name everywhere else. And maybe it's fine because only devs use `cilium monitor` nowadays?

ldelossa replied:
Ah I see, if its just for Cilium monitor output then I think that's cool. I'll make the change.

ldelossa replied:
Fixed in latest push

pchaigno commented on bpf/bpf_host.c:
```diff
@@ -1308,12 +1308,46 @@ int cil_from_host(struct __ctx_buff *ctx)
+	struct iphdr __maybe_unused *ipv4;
+	void __maybe_unused *data, *data_end = NULL;
+
+	/* only support vxlan for now */
```
Could we have a good-first-issue for GENEVE support? It's easy to do and the sort of small feature gap that makes for an awkward user experience.

ldelossa replied:
Yup, Joe pointed this out as well, we can open this ticket after merge.

pchaigno commented on bpf/lib/encrypt.h:
```diff
@@ -132,6 +133,84 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+	if (eth_load_saddr(ctx, &dst_mac, 0) != 0)
+		return DROP_INVALID;
+	if (eth_store_daddr(ctx, &dst_mac, 0) != 0)
+		return DROP_WRITE_ERROR;
```
Did you check this is needed? It might be but I'm also wondering if we're not doing these rewrites too often.

ldelossa replied:
Definitely needed. The ingress packet path will drop the redirect if the MAC does not correspond to the receiving device. It occurs in https://elixir.bootlin.com/linux/latest/source/net/ethernet/eth.c#L155

ldelossa replied:
Going to resolve for now, since in my testing this is required or else we drop the packet with a martian source. If you feel otherwise feel free to unresolve and ping me.

pchaigno replied:
Ah, I see. That's what sets `skb->pkt_type`. I thought maybe it was avoidable by setting the packet type ourselves, but it's the other way around :+1:

julianwiedmann commented on bpf/lib/tunnel.h:
```diff
@@ -5,6 +5,11 @@
+#include <linux/ip.h>
+#include <linux/udp.h>
+#include <linux/if_ether.h>
+#include "lib/csum.h"
```
Just to note - as mentioned offline, I had forked off `tunnel.h` to be the "simple" header that just contains definitions and trivial helpers. Your additions *might* be better placed in `encap.h` for the time being. But no biggie.

ldelossa replied:
Hm, I'm definitely not opposed to moving the vxlan helper functions somewhere else? Maybe their own header file? I think `lib/csum.h` is pretty light weight right now. 

While just a note, if we want to handle this better, maybe a new file for the vxlan helper functions is warranted?

julianwiedmann replied:
> maybe a new file for the vxlan helper functions is warranted

works for me :)

ldelossa replied:
Fixed in latest push, resolving

julianwiedmann replied:
Do we still need these additional includes then? :thinking:

ldelossa replied:
Believe so, those include are providing us net-protocol headers and macros (ipv4, udp, mac, etc..) and the csum is for rewriting checksum.

julianwiedmann replied:
So `vxlan.h` needs them. But `tunnel.h` has no diff, except adding these includes ...?

ldelossa replied:
Ah! My bad I thought we were looking at vxlan.h 😅. 

Yup, ill remove these!

ldelossa replied:
Okay, backed out any touches to tunnel.h, tunnel.h should indeed have no diff at all in this PR.

julianwiedmann reviewed (changes requested):
Had a closer look - otherwise lgtm, thank you! Just one nit below, and blocking on the `ENABLE_ENCRYPTED_OVERLAY` change.

ldelossa committed:
ipsec: add encrypted overlay flags

This commits and both the agent and datapath flag required to enable the
"Encrypted Overlay" feature.

The datapath will use ENABLE_ENCRYPTED_OVERLAY feature flag.
The agent will use "encryption.ipsec.encryptOverlay"

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa commented on pkg/datapath/linux/config/config.go:
```diff
@@ -279,6 +279,10 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 	if option.Config.EnableIPSec {
 		cDefinesMap["ENABLE_IPSEC"] = "1"
+
+		if option.Config.EnableIPSecEncryptedOverlay {
```
this gets nested because we'll simply ignore it if IPSec is not enabled.

julianwiedmann commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -764,6 +764,8 @@ encryption:
     keyRotationDuration: "5m"
+    # -- Enable IPSec encrypted overlay
+    encryptedOverlay: true
   wireguard:
```
Hmm, why do we default to `true` here? I would tend towards being conservative, and starting with an explicit opt-in.

ldelossa replied:
hmm, 

You are right, but I'm not sure when this tmpl is used, our defaults inside Cilium are

```
	// Enable IPSec encrypted overlay
	//
	// This feature will encrypt overlay traffic before it leaves the cluster.
	EnableIPSecEncryptedOverlay = false
```

Anyway, it should be false but now I'm a bit curious.

ldelossa replied:
fixed in latest push, resolving.

julianwiedmann commented on daemon/cmd/daemon_main.go:
```diff
@@ -1349,6 +1352,16 @@ func initEnv(vp *viper.Viper) {
+	if option.Config.EnableIPSecEncryptedOverlay && !option.Config.EnableIPSec {
+		log.Warn("IPSec encrypted overlay is enabled but IPSec is not. Ignoring option.")
+	}
+
```
We would also want to check that the TunnelProtocol is VXLAN, right?

ldelossa replied:
I didn't see the flag for this, actually, I think this particular setting is jammed inside a cell and not accessible from here, IIUC. Can you show me how I can sus out the TunnelProtocol? I did not see it anywhere in `option.Config`

ldelossa replied:
Fixed this in latest push. Winds up doing validation in `newDaemon` provides us access to the TunnelConfig object on the daemon's hive parameters.

julianwiedmann commented on pkg/option/config.go:
```diff
@@ -711,6 +711,12 @@ const (
+	// the EncryptedOverlay feature.
+	//
+	// This feature will encrypt overlay traffic before it leaves the cluster.
+	EnableIPSecEncrytpedOverlay = "enable-ipsec-encrypted-overlay"
```
```suggestion
	EnableIPSecEncryptedOverlay = "enable-ipsec-encrypted-overlay"
```

ldelossa replied:
I knew there'd be at least one :laughing:

ldelossa replied:
fixed in latest push

julianwiedmann commented on pkg/datapath/linux/config/config.go:
```diff
@@ -278,6 +279,10 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 		cDefinesMap["ENABLE_IPSEC"] = "1"
+
+		if option.Config.EnableIPSecEncryptedOverlay {
+			cDefinesMap["ENABLE_ENCRYPTED_OVERLAY"] = "1"
+		}
```
We don't have any compile coverage for the affected paths atm, right? Could you please wire up `bpf/Makefile`?

ldelossa replied:
fixed in latest push.

julianwiedmann reviewed (commented):
Just some minor comments, otherwise lgtm!

julianwiedmann reviewed (approved):
looks good, thank you!

pchaigno commented on bpf/Makefile:
```diff
@@ -92,11 +92,11 @@ HOST_OPTIONS = $(LXC_OPTIONS) \
+	-DDISABLE_LOOPBACK_LB:-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DPOLICY_VERDICT_NOTIFY:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_DSR:-DENABLE_DSR_HYBRID:-DENABLE_PREFILTER:-DENABLE_SESSION_AFFINITY:-DENABLE_HOST_FIREWALL:-DENABLE_ICMP_RULE:-DENABLE_SRV6:-DENABLE_MULTICAST:-DENCRYPTED_OVERLAY: \
+	-DDISABLE_LOOPBACK_LB:-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DPOLICY_VERDICT_NOTIFY:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_DSR:-DENABLE_DSR_HYBRID:-DENABLE_PREFILTER:-DENABLE_SESSION_AFFINITY:-DENABLE_HOST_FIREWALL:-DENABLE_ICMP_RULE:-DENABLE_SRV6:-DENABLE_SRV6_SRH_ENCAP:-DENABLE_SCTP:-DENABLE_MULTICAST:-DENCRYPTED_OVERLAY: \
+	-DDISABLE_LOOPBACK_LB:-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DPOLICY_VERDICT_NOTIFY:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_DSR:-DENABLE_DSR_HYBRID:-DENABLE_PREFILTER:-DENABLE_SESSION_AFFINITY:-DENABLE_HOST_FIREWALL:-DENABLE_ICMP_RULE:-DENABLE_SRV6:-DENABLE_SRV6_SRH_ENCAP:-DENABLE_SCTP:-DENABLE_VTEP:-DENABLE_MULTICAST:-DENCRYPTED_OVERLAY: \
+	-DDISABLE_LOOPBACK_LB:-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DPOLICY_VERDICT_NOTIFY:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_DSR:-DDSR_ENCAP_MODE:-DDSR_ENCAP_GENEVE:-DENABLE_PREFILTER:-DENABLE_SESSION_AFFINITY:-DENABLE_HOST_FIREWALL:-DENABLE_ICMP_RULE:-DENABLE_SRV6:-DENABLE_SRV6_SRH_ENCAP:-DENCRYPTED_OVERLAY:-DENABLE_SCTP:-DENABLE_VTEP:-DENABLE_HIGH_SCALE_IPCACHE:-DDSR_ENC ...
```
We also need to add this to bpf/complexity-tests were appropriate to cover complexity issues. It can be a follow up, but we shouldn't wait too much or we risk having complexity issues.

ldelossa replied:
Will start testing this after this merge.
